### ISSUE TITLE
Magento 2.2.4 not sending from sales sender

### ISSUE THREAD

**Issue by versdivers** [magento/magento2#16355]
### Preconditions
1.  Magento 2.2.4
2.  PhP 7.1
3.  Apache

### Steps to reproduce
1.  Set general store email addresses
2.  Set store specific email senders in Sales Emails
3. Send order/shipment/creditnota/invoice mail to customer

### Expected result
Email is sent from store email address

### Actual result
It sends from the mail server default mailadress

### Why

In vendor **\magento\module-sales\Model\Order\Email\SenderBuilder.php** the `configureEmailTemplate()` function as been changed.

You can see that 
`        $this->transportBuilder->setFrom($this->identityContainer->getEmailIdentity());
`
has been changed to 
 
```
        $this->transportBuilderByStore->setFromByStore(
            $this->identityContainer->getEmailIdentity(),
            $this->identityContainer->getStore()->getId()
        );
```

With further inspection it on first glance it returns the same result .

Eventually the from gets formatted within the `_formatAddress `function within **Zend/Mail.php** and used within the same class at setFrom function which has 
`        $this->_storeHeader('From', $this->_formatAddress($email, $name), true);`

The 'From' header can have something like _Test Mail <test@mail.com>_ or just _test@mail.com_. But the function itself returns just _Test Mail_ with no email.

And that is because the `sprintf `function appearently seems to have issues with `<>` signs. Note sure why.

So i changed
```
            $encodedName = $this->_encodeHeader($name);
            if ($encodedName === $name  &&  strcspn($name, '()<>[]:;@\\,.') != strlen($name)) {
                $format = '"%s" <%s>';
            } else {
                $format = '%s <%s>';
            }
            return sprintf($format, $encodedName, $email);
```

to

```
            $encodedName = $this->_encodeHeader($name);
            if ($encodedName === $name  &&  strcspn($name, '()<>[]:;@\\,.') != strlen($name)) {
                $format = '"%1$s" %3$s%2$s%4$s';
            } else {
                $format = '%1$s %3$s%2$s%4$s';
            }
            return sprintf($format, $encodedName, $email , '<' , '>');
```

and now it does return  _Test Mail <test@mail.com>_. So that is fixed;

And now i am stuck. 

`this->transportBuilder->setFrom($this->identityContainer->getEmailIdentity()); `

-->sends what i need.

```
      $this->transportBuilderByStore->setFromByStore(
            $this->identityContainer->getEmailIdentity(),
            $this->identityContainer->getStore()->getId()
        );
```

-->sends from my SMTP servers default emailaddress.

It seems like the headers are reset somewhere.

**Comment by magento-engcom-team** [comment]
Hi @versdivers. Thank you for your report. 
To help us process this issue please make sure that you provided the following information:
- [ ] Summary of the issue
- [ ] Information on your environment
- [ ] Steps to reproduce
- [ ] Expected and actual results

Please make sure that the issue is reproducible on the vanilla Magento instance following **Steps to reproduce**. To deploy vanilla Magento instance on our environment, please, add a comment to the issue: 

`@magento-engcom-team give me {$VERSION} instance` 

where `{$VERSION}` is [version tags](https://github.com/magento/magento2/tags) (starting from 2.2.0+) or develop branches (2.2-develop +). 
For more details, please, review the [Magento Contributor Assistant](https://github.com/magento/magento2/wiki/Magento-Contributor-Assistant) documentation.

@versdivers do you confirm that you was able to reproduce the issue on vanilla Magento instance following steps to reproduce?
- [ ] yes
- [ ] no

**Comment by versdivers** [comment]
@magento-engcom-team  yes

**Comment by versdivers** [comment]
It seems that TransportBuilder and transportBuilderByStore work in different instances of the messageinterface

**Comment by versdivers** [comment]
i also do not get why transportBuilderByStore  as been created. I see this just as bad coding tbh. We are back to creating new classes for simple adjustments.

Fix:
**vendor/magento/module-sales/Model/Order/Email/SenderBuilder.php**

replace 
```
        $this->transportBuilderByStore->setFromByStore(
            $this->identityContainer->getEmailIdentity(),
            $this->identityContainer->getStore()->getId()
        );
```

with

`$this->transportBuilder->setFrom($this->identityContainer->getEmailIdentity(), $this->identityContainer->getStore()->getId());
`
**vendor/magento/framework/Mail/Template/TransportBuilder.php**
replace

```
    /**
     * Set mail from address
     *
     * @param string|array $from
     * @return $this
     */
    public function setFrom($from)
    {
        $result = $this->_senderResolver->resolve($from);
        $this->message->setFrom($result['email'], $result['name']);
        return $this;
    }
```

with

```
    /**
     * Set mail from address
     *
     * @param string|array $from
     * @return $this
     */
    public function setFrom($from, $store = null)
    {
        $result = $this->_senderResolver->resolve($from, $store);
        $this->message->setFrom($result['email'], $result['name']);
        return $this;
    }
```

and just like that we do not need the TransportBuilderByStore anymore and it works as expected.

**Comment by kingyond** [comment]
@versdivers this issue same as #16231.
The bug cause by amazon payment module. It disable Magento\Framework\Mail\MessageInterface as singleton object.  And the make TransportBuilderByStore create a new MessageInterface which different from the on create by TransportBuilder.
open this file
vendor/amzn/amazon-pay-and-login-magento-2-module/src/Payment/etc/di.xml
find this line #63
`<type name="Magento\Framework\Mail\MessageInterface" shared="false" />`
remove the shared string
`<type name="Magento\Framework\Mail\MessageInterface" />`
save it.
after that, remove generated data and compile the di again. this should fix the bug.

**Comment by gwharton** [comment]
@versdivers Your comment above should have been the right solution in my opinion. But was turned down on review as it meant making a change to a class marked as an API.

Instead we have the bodged solution that is currently in place, which still has bugs, e.g you can no longer send multiple emails at the same time, only the first will succeed, the rest will exception and fail.

**Comment by versdivers** [comment]
I will keep my solution in place. On update i hope Magento found a way to remove all the bugs with this but for those who run into this problem i can comfirm that this solution is working great on a live website with multistore.

**Comment by gwharton** [comment]
Duplicate of #14952

**Comment by rishabhchd19** [comment]
Resolved Problem Disabling AMAZON PAYMENT.

**Comment by mjoraid** [comment]
1- I agree with others. Disabling Amazon Payment does fix the issue. I disabled payment, login, and core just for the sake of it and then enabled them again.
2- +kingyond solutions did fix the issue for me. Now order email is sent From as the I specified in my Admin - Sales - Sales Emails 

```
vendor/amzn/amazon-pay-and-login-magento-2-module/src/Payment/etc/di.xml
find this line #63
<type name="Magento\Framework\Mail\MessageInterface" shared="false" />
remove the shared string
<type name="Magento\Framework\Mail\MessageInterface" />
save it.
```

Now I'm not sure what are the implications of removing 

` share="false"`

**Comment by ihor-sviziev** [comment]
Related issue: https://github.com/magento/magento2/issues/14952

**Comment by magento-engcom-team** [comment]
Hi @versdivers. Thank you for your report. 
The issue has been fixed in magento/magento2#18471 by @gwharton in 2.3-develop branch
Related commit(s):
 - [ad50297e7bf4ea39db62f0969579a1ca8eb5f1a3](https://github.com/magento/magento2/commit/ad50297e7bf4ea39db62f0969579a1ca8eb5f1a3)
 - [e16ff74e4f5f24a4a2530b6aced6ef50032515e0](https://github.com/magento/magento2/commit/e16ff74e4f5f24a4a2530b6aced6ef50032515e0)
 - [621d2cfeaf8e0a15e592cdc001c856385b3cbd83](https://github.com/magento/magento2/commit/621d2cfeaf8e0a15e592cdc001c856385b3cbd83)
 - [b6f9e48289119117df230b6578ac9ff3d69e4996](https://github.com/magento/magento2/commit/b6f9e48289119117df230b6578ac9ff3d69e4996)
 - [0dc85785e522e2be8c72dfeaee7e9f5054195034](https://github.com/magento/magento2/commit/0dc85785e522e2be8c72dfeaee7e9f5054195034)
 - [122019c953037df74ef97245a42b274eae2e6639](https://github.com/magento/magento2/commit/122019c953037df74ef97245a42b274eae2e6639)
 - [193c6c2f6facf7276c1c8858d29ea3b847062ef5](https://github.com/magento/magento2/commit/193c6c2f6facf7276c1c8858d29ea3b847062ef5)
 - [bd0aa57cc837918335015d051f08cf380dc186d0](https://github.com/magento/magento2/commit/bd0aa57cc837918335015d051f08cf380dc186d0)
 - [523b0464966f81f1f01aa13073092bfddf363219](https://github.com/magento/magento2/commit/523b0464966f81f1f01aa13073092bfddf363219)
 - [424929fe9aee1005601cc2c9c7ad0090741be723](https://github.com/magento/magento2/commit/424929fe9aee1005601cc2c9c7ad0090741be723)
 - [c32ca7a0d58cbc57276349b26eb5ecd38451fd2d](https://github.com/magento/magento2/commit/c32ca7a0d58cbc57276349b26eb5ecd38451fd2d)
 - [e39ce7a3f2a7e6ee134cff9a097b7321756c32cf](https://github.com/magento/magento2/commit/e39ce7a3f2a7e6ee134cff9a097b7321756c32cf)
 - [99621b32f8c7d3d12f83db7e67f5f1c1ef3e7f31](https://github.com/magento/magento2/commit/99621b32f8c7d3d12f83db7e67f5f1c1ef3e7f31)
 - [f171c9d6ab035439ffb6374ec14dd8d51123dcd2](https://github.com/magento/magento2/commit/f171c9d6ab035439ffb6374ec14dd8d51123dcd2)

The fix will be available with the upcoming 2.3.1 release.

**Comment by pratikhmehta** [comment]
> i also do not get why transportBuilderByStore as been created. I see this just as bad coding tbh. We are back to creating new classes for simple adjustments.
> 
> Fix:
> **vendor/magento/module-sales/Model/Order/Email/SenderBuilder.php**
> 
> replace
> 
> ```
>         $this->transportBuilderByStore->setFromByStore(
>             $this->identityContainer->getEmailIdentity(),
>             $this->identityContainer->getStore()->getId()
>         );
> ```
> with
> 
> `$this->transportBuilder->setFrom($this->identityContainer->getEmailIdentity(), $this->identityContainer->getStore()->getId()); `
> **vendor/magento/framework/Mail/Template/TransportBuilder.php**
> replace
> 
> ```
>     /**
>      * Set mail from address
>      *
>      * @param string|array $from
>      * @return $this
>      */
>     public function setFrom($from)
>     {
>         $result = $this->_senderResolver->resolve($from);
>         $this->message->setFrom($result['email'], $result['name']);
>         return $this;
>     }
> ```
> with
> 
> ```
>     /**
>      * Set mail from address
>      *
>      * @param string|array $from
>      * @return $this
>      */
>     public function setFrom($from, $store = null)
>     {
>         $result = $this->_senderResolver->resolve($from, $store);
>         $this->message->setFrom($result['email'], $result['name']);
>         return $this;
>     }
> ```
> and just like that we do not need the TransportBuilderByStore anymore and it works as expected.

Thank You so much.
Also Working in Magento 2.2.7.

**Comment by gwharton** [comment]
@pratikhmehta totally agree. The transportbuilderbystore class has already been deprecated in 2.3 and soon will be in 2.2.

**Comment by pratikhmehta** [comment]
> @pratikhmehta totally agree. The transportbuilderbystore class has already been deprecated in 2.3 and soon will be in 2.2.

But early I used Magento 2.2.4 that time i did not get any issue like this. After upgrade version Magento 2.2.7 Issue is coming. 
But after used [solution above](https://github.com/magento/magento2/issues/16355#issue-335073411) solved issue.

**Comment by gwharton** [comment]
Yes, the problem was introduced to fix a separate issue in 2.2.5, however the fix had unexpected consequences, including this one.

You will see in PR #18472 that a fix very much along the lines of your comments above, is going to be merged very soon into 2.2 and hopefully this will be available in time for release 2.2.8.

Like I said, the fix has already been applied to 2.3 in PR #18471

**Comment by magento-engcom-team** [comment]
Hi @versdivers. Thank you for your report. 
The issue has been fixed in magento/magento2#18472 by @gwharton in 2.2-develop branch
Related commit(s):
 - [82a635b3a4eb0fd0e2f030d82495ab717ee11db5](https://github.com/magento/magento2/commit/82a635b3a4eb0fd0e2f030d82495ab717ee11db5)
 - [16afc0962463261e4464456b9bdaad4e8f3be924](https://github.com/magento/magento2/commit/16afc0962463261e4464456b9bdaad4e8f3be924)
 - [47c5f72c9196dc77b50db3c36ec99719a69a18a0](https://github.com/magento/magento2/commit/47c5f72c9196dc77b50db3c36ec99719a69a18a0)
 - [8d7a4513ddf4b759a6c6a0d2ace04aa714a396df](https://github.com/magento/magento2/commit/8d7a4513ddf4b759a6c6a0d2ace04aa714a396df)
 - [a7cd527a6896c3a2f204e2771b92496ef4089f4e](https://github.com/magento/magento2/commit/a7cd527a6896c3a2f204e2771b92496ef4089f4e)
 - [491d7f133591c64346cb54c2392d0ef212333b1a](https://github.com/magento/magento2/commit/491d7f133591c64346cb54c2392d0ef212333b1a)
 - [490e5d22074224ec1965a391abfffc379c31d0ee](https://github.com/magento/magento2/commit/490e5d22074224ec1965a391abfffc379c31d0ee)
 - [b19b3ed195b870f16a00c673e435b5b9b8184227](https://github.com/magento/magento2/commit/b19b3ed195b870f16a00c673e435b5b9b8184227)
 - [9f34fd631a55ded6d8a458d219edfc23c1bc5f9b](https://github.com/magento/magento2/commit/9f34fd631a55ded6d8a458d219edfc23c1bc5f9b)
 - [66478ec59d6cbf77d538f2ea112f32997e491bf8](https://github.com/magento/magento2/commit/66478ec59d6cbf77d538f2ea112f32997e491bf8)
 - [cc34f00113e1e644e92002f82d34c4bf36aa0e26](https://github.com/magento/magento2/commit/cc34f00113e1e644e92002f82d34c4bf36aa0e26)
 - [fb572d887f0ce156c8f3ac149de7b8f06835623d](https://github.com/magento/magento2/commit/fb572d887f0ce156c8f3ac149de7b8f06835623d)
 - [59a4e57a0c36671ff84e8398f413253610a512ac](https://github.com/magento/magento2/commit/59a4e57a0c36671ff84e8398f413253610a512ac)
 - [a8bcb8bb8442a649bf1e62400ca6f147023e53c6](https://github.com/magento/magento2/commit/a8bcb8bb8442a649bf1e62400ca6f147023e53c6)
 - [573cf7a9fee7c2d8ca9c1c25fe909c93ccf06c49](https://github.com/magento/magento2/commit/573cf7a9fee7c2d8ca9c1c25fe909c93ccf06c49)
 - [ef635bf6b8ecf67c729deb38df05c81b9e5b5359](https://github.com/magento/magento2/commit/ef635bf6b8ecf67c729deb38df05c81b9e5b5359)
 - [daa8996bbda2e7ca938a7e3a3e8afc40502cd98f](https://github.com/magento/magento2/commit/daa8996bbda2e7ca938a7e3a3e8afc40502cd98f)

The fix will be available with the upcoming 2.2.9 release.

**Comment by CompactCodeEU** [comment]
@magento-engcom-team You already said that this would be fixed in 2.3. I do not think that there is an upcoming 2.2.9 release anymore. That version has already been released. Do you mean there is a patch for 2.2.9? We still changed it in 2.3.0 though for our customers since the fix was not implemented.

**Comment by ihor-sviziev** [comment]
Hi @CompactCodeEU,
Releases in 2.3.x and 2.2.x lines are going together, so this fix should be included in upcoming 2.3.1 and 2.2.9 releases, as @magento-engcom-team said
